### DISCUSSION TITLE
How to build (or get) pure-python wheel

### DISCUSSION THREAD

**Discussion by castedo** [jelmer/dulwich#1397] [Q&A] [answered]
It looks like there is no pure python wheel distributed to PyPI. When I try to load dulwich from the [Pyodide REPL](https://pyodide.org/en/stable/console.html)
```
Welcome to the Pyodide 0.26.3 terminal emulator 🐍
Python 3.12.1 (main, Oct 19 2024 08:33:26) on WebAssembly/Emscripten
Type "help", "copyright", "credits" or "license" for more information.
>>> import micropip
>>> await micropip.install("dulwich")
>>> micropip.list()
Name      | Version  | Source
--------- | -------- | -------
packaging | 23.2     | pyodide
micropip  | 0.6.0    | pyodide
dulwich   | 0.20.1   | pypi
certifi   | 2024.2.2 | pyodide
urllib3   | 2.2.1    | pyodide
>>> import dulwich
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ModuleNotFoundError: No module named 'dulwich'
>>> 
```

I'm guessing this is because there is no pure-python wheel on PyPI.

I'm trying to build a pure python wheel from the source repository now but I'm having trouble figuring out which build tool I should use. Should I use `python3 -m build` to build a pure-python wheel?

**Comment by castedo** [comment]
I think this might just be a problem with the filename of the wheel file generated. When I look inside the wheel while (using `unzip -l` I don't see any binary files like I do for the wheels on PyPI.

I'm currently trying to use `python3 -m pip wheel ...` create a pure-python wheel from the source tar ball.

**Reply by castedo** [reply]
Yep, it's a pure-python wheel that is created. I'm able to import it in the Pyodide REPL hosted locally when I rename the wheel file. I am able to follow all the steps in the [tutorial](https://dulwich.io/docs/tutorial/repo.html) all from within the Pyodide REPL hosted locally along with a wheel file that I have manually renamed to end in `py3-none-any.whl`.

**Comment by castedo** [comment]
So there are two separate issues here:

1) maybe it is worth putting a pure-python wheel up on PyPI so that Pyodide usage is less confusing
2) what is the proper way to use `python3 -m pip wheel ...` or similar to create a pure-python wheel from the source tar ball?

I'm confused why `python3 -m pip wheel ...` is creating a wheel with a filename that sounds like it's binary wheel even though I've tried passing `--no-binary` and the contents have nothing binary inside.

**Reply by jelmer** [reply, marked as answer]
See https://github.com/jelmer/dulwich/pull/1398

**Reply by castedo** [reply]
Nice! That was fast! :smile:

Oooooo, I see `PURE=true python -m build --wheel` is the trick. In the short-term I'll use `PURE=true`.

So long-term this pure-python wheel will eventually show up on PyPI?

**Reply by jelmer** [reply]
Yeah, we'll have pure-python wheels in pypi in the future.
